sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UnchangedLocalVariablesShouldBeConst.cs

```diff
@@ -0,0 +1,189 @@
+                    {
+                        return;
+                    }
+
+                    localDeclaration?.Declaration?.Variables
+                        .Where(v => v?.Identifier != null)
+                        .Select(v => new { Syntax = v, Symbol = c.SemanticModel.GetDeclaredSymbol(v) })
+                        .Where(ss => ss.Symbol != null)
+                        .Where(ss => IsInitializedWithCompatibleConstant(ss.Syntax, c.SemanticModel, declaredType))
+                        .Where(ss => !HasMutableUsagesInMethod(ss.Syntax, ss.Symbol, c.SemanticModel))
+                        .ToList()
+                        .ForEach(ss => Report(ss.Syntax, c));
+                },
+                SyntaxKind.LocalDeclarationStatement);
+        }
+
+        private static DeclarationType FindDeclarationType(LocalDeclarationStatementSyntax localDeclaration,
+            SemanticModel semanticModel)
+        {
+            var declaredTypeSyntax = localDeclaration.Declaration?.Type;
+
+            if (declaredTypeSyntax == null)
+            {
+                return DeclarationType.Unknown;
+            }
+
+            var declaredType = semanticModel.GetTypeInfo(declaredTypeSyntax).Type;
+            if (declaredType == null)
+            {
+                return DeclarationType.Unknown;
+            }
+
+            if (declaredType.Is(KnownType.System_String))
+            {
+                return DeclarationType.String;
+            }
+
+            return declaredType.IsValueType
+                ? DeclarationType.Value
+                : DeclarationType.Reference;
+        }
+
+
+        private static bool IsInitializedWithCompatibleConstant(VariableDeclaratorSyntax variableDeclarator,
+            SemanticModel semanticModel, DeclarationType declarationType)
+        {
+            var initializer = variableDeclarator?.Initializer?.Value;
+            if (initializer == null)
+            {
+                return false;
+            }
+            var constantValueContainer = semanticModel.GetConstantValue(initializer);
+            if (!constantValueContainer.HasValue)
+            {
+                return false;
+            }
+
+            var constantValue = constantValueContainer.Value;
+
+            if (constantValue is string)
+            {
+                return declarationType == DeclarationType.String;
+            }
+
+            if (constantValue is ValueType)
+            {
+                return declarationType == DeclarationType.Value;
+            }
+
+            return declarationType == DeclarationType.Reference ||
+                   declarationType == DeclarationType.String;
+        }
+
+        private static bool HasMutableUsagesInMethod(VariableDeclaratorSyntax parameter, ISymbol parameterSymbol,
+            SemanticModel semanticModel)
+        {
+            var methodSyntax = parameter.Parent?.AncestorsAndSelf()?.FirstOrDefault(IsMethodLike);
```

Comment:
Why `Parent?.AncestorsAndSelf()` instead of `Ancestors()`?

Reply:
That would be too simple ;)

Reply:
I like simple :)